bundle/src/main/java/com/adobe/acs/commons/models/injectors/impl/ChildRequestInjector.java

```diff
@@ -0,0 +1,149 @@
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.models.spi.DisposalCallbackRegistry;
+import org.apache.sling.models.spi.Injector;
+import org.osgi.framework.Constants;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.AnnotatedElement;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Injector implementation for `@ChildRequest`
+ * <p>
+ * If the current adaptable is a `SlingHttpServletRequest`, injects a single
+ * or list of `SlingHttpServletRequest` instance(s) referencing the child
+ * resource path(s).
+ * <p>
+ * If the current adaptable is a `Resource`, injects a single or list of
+ * `Resource` child resources similar to the `@ChildResource injector.
+ */
+@Component(
+        service = {Injector.class},
+        property = {
+                Constants.SERVICE_RANKING + "=3000"
+        }
+)
+public class ChildRequestInjector implements Injector {
+    private static final Logger logger = LoggerFactory.getLogger(ChildRequestInjector.class);
+
+    public String getName() {
+        return "child-requests";
+    }
+
+    public Object getValue(Object adaptable, String name, Type declaredType, AnnotatedElement element, DisposalCallbackRegistry callbackRegistry) {
+        // sanity check
+        if (element.getAnnotation(ChildRequest.class) == null) {
+            return null;
+        }
+
+        if (adaptable instanceof SlingHttpServletRequest) {
+            logger.trace("Injecting '{}' from request", name);
+            Class injectionType = null;
+            if (declaredType instanceof Class) {
+                injectionType = (Class) declaredType;
+            } else if (this.isDeclaredTypeCollection(declaredType)) {
+                injectionType = this.getActualType((ParameterizedType) declaredType);
+            }
+            if (injectionType != null) {
+                SlingHttpServletRequest request = (SlingHttpServletRequest) adaptable;
+                if (Resource.class.isAssignableFrom(injectionType)) {
+                    logger.trace("Injected object type is resource, so injecting as resource rather than request");
+                    return getValueForResource(request.getResource(), name, declaredType);
+                } else {
+                    return getValueForRequest(request, name, declaredType);
+                }
+            }
+        } else if (adaptable instanceof Resource) {
+            logger.trace("Injecting '{}' from resource (request not available)", name);
+            return getValueForResource((Resource) adaptable, name, declaredType);
+        }
+        return null;
+    }
+
+    private Object getValueForRequest(SlingHttpServletRequest request, String name, Type declaredType) {
```

Comment:
This and `getValueForResource` looks identical except for 1 line; could this be abstracted to reduce code dupe?

Reply:
through this magics of java 8, this is resolved